[fix] at::view

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23239 [fix] at::view**

accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

jerryzh168 committed:
[fix] at::view

accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

jerryzh168 committed:
Update on "[fix] at::view"


accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

jerryzh168 committed:
Update on "[fix] at::view"


accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

gchanan commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -909,7 +909,28 @@ Tensor view(const Tensor& self, IntArrayRef size) {
+    self_ = at::_empty_affine_quantized({0}, self.options(), self.q_scale(),
+                                        self.q_zero_point());
+  } else if (self.is_cuda()) {
+    // NOTE: This path of constructing the Tensor directly with the viewed
```
where is this coming from?  shouldn't empty with the options below create a tensor on the correct device?

Also, not necessary for this PR, but we really need an "alias" that works across all tensor types.

jerryzh168 replied:
This comes from THCTensor_(view), should we remove it?

jerryzh168 replied:
you mean `at::alias`, there are a some legacy TH functions still used in aten, I think I can move them to aten if quantized Tensor needs them in the future.

jerryzh168 committed:
Update on "[fix] at::view"


accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

jerryzh168 commented on aten/src/THC/generic/THCTensor.cpp:
```diff
@@ -237,7 +210,7 @@ THCTensor *THCTensor_(newFoldBatchDim)(THCState *state, THCTensor *input) {
     new_size[i - 1] = THCTensor_(size)(state, input, i);
   }
-  THCTensor *output = THCTensor_(newView)(state, input, new_size);
+  THCTensor *output = at::native::view(THTensor_wrap(input), new_size).unsafeGetTensorImpl();
```
@gchanan is this OK?

dzhulgakov replied:
Looks ok to me, but I'm not a TH expert

gchanan replied:
lgtm, you can look at the generated code for this boundary (which calls checked_tensor_unwrap) and it does the same thing

dzhulgakov commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -909,7 +909,15 @@ Tensor view(const Tensor& self, IntArrayRef size) {
+    self_ = at::_empty_affine_quantized({0}, self.options(), self.q_scale(),
+                                        self.q_zero_point());
+  } else {
+    self_ = at::empty({0}, self.options());
```
this is still more expensive - we create an empty storage here and the go on to replace it. You probably need to drop down one level lower and create tensor with explicitly provided storage instead

jerryzh168 committed:
Update on "[fix] at::view"


accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

jerryzh168 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -909,9 +909,21 @@ Tensor view(const Tensor& self, IntArrayRef size) {
+    auto impl = c10::make_intrusive<QTensorImpl>(Storage(self.storage()), self.type_id(),
+                    get_qtensorimpl(self)->quantizer());
+    impl->set_storage_offset(self.storage_offset());
+    impl->set_sizes_and_strides(inferred_size, stride_value);
```
cc @VitalyFedyunin do we need to inlcude memoryformat here?

VitalyFedyunin replied:
No. It is taken from the strides information.

jerryzh168 committed:
Update on "[fix] at::view"


accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

gchanan reviewed (commented):
is it possible to make `alias` work with quantized tensors?  That way you could just change your `clone` call to `alias` and you wouldn't need the logic in multiple places.

jerryzh168 commented:
@gchanan this line doesn't raise after the change: https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L3473
do you know why?

gchanan commented:
yes, it's because TH stuff needs to be explicitly enabled for bfloat16, but native functions doesn't.

The correct thing to do here is to get rid of that entire if block as this is an improvement.

CC @izdeby: this is why you should comment these asserts, so people know it's a _good_ thing if they remove them, not a bad thing.

jerryzh168 committed:
Update on "[fix] at::view"


accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

gchanan commented on aten/src/TH/generic/THTensor.cpp:
```diff
@@ -174,41 +159,25 @@ THTensor *THTensor_(newContiguous)(THTensor *self)
 THTensor *THTensor_(newSelect)(THTensor *tensor, int dimension_, int64_t sliceIndex_)
 {
-  THTensor *self = THTensor_(newWithTensor)(tensor);
+  THTensor *self = at::native::alias(THTensor_wrap(tensor)).unsafeGetTensorImpl();
```
nit: I think it's cleaner to just leave `THTensor_(newWithTensor)` and have it do this internally.

When there's only 1 call like with newView I think the tradeoff is in the other direction.

jerryzh168 replied:
don't we want to remove `THTensor_(newWithTensor)`?

gchanan replied:
we definitely want to remove the binding and the separate implementation.

We want to remove the method as well, but I don't think the aesthetics are worth it at this point: you'd have to move most of the callers to make it not ugly IMO.

gchanan commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -909,9 +909,27 @@ Tensor view(const Tensor& self, IntArrayRef size) {
+                    get_qtensorimpl(self)->quantizer());
+    self_ = Tensor(impl);
+  } else {
+    auto impl = c10::make_intrusive<TensorImpl>(Storage(self.storage()),
```
this doesn't look right -- in the call you are replacing, the sizes and strides are set correctly.  This doesn't matter for `view` because you never look at the alias before you reset it, but this is wrong for the other users of alias.

jerryzh168 replied:
oh right.

gchanan commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -909,9 +909,27 @@ Tensor view(const Tensor& self, IntArrayRef size) {
-  auto self_ = self.clone();
-  self_.set_(self.storage(), self.storage_offset(), inferred_size,
-             stride_value);
+  auto self_ = self.alias();
```
I can't comment on the above, but this also looks wrong:
```
  if (self.sizes() == inferred_size) {
    return self;
}
```

I don't know if you can induce an error from the python frontend because we have a bunch of extra code for Variable-scope views, but in non-Variable mode, doesn't this mean that:
```
Tensor x = at::randn(2,3);
y = x.view(2,3);
x.resize_(3,2);
// y will be resized even though it's a view!
```

can you verify this / add a test?

jerryzh168 replied:
yeah this looks wrong.

jerryzh168 replied:
just added some tests, looks like it `y` didn't got resized.
maybe it's because we pass in const Tensor& and when we return `self` it got copied? but it should just copy the `intrusive_ptr` rather than `TensorImpl`?

I saw Tensor has default copy constructor:
`Tensor(const Tensor&) = default;`, does it copy the underlying `impl_`: `c10::intrusive_ptr<TensorImpl, UndefinedTensorImpl> impl_;`?

gchanan replied:
`Tensor` doesn't matter, what matters if there is a different `TensorImpl` or not.  I think you'd have to write an `at::` test to find that bug, we have tests like that in `/aten/src/ATen/test/`.

jerryzh168 replied:
yeah testing tensor directly will catch the bug.

jerryzh168 committed:
Update on "[fix] at::view"


accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

gchanan commented on test/test_torch.py:
```diff
@@ -8410,6 +8407,12 @@ def _test_view(self, cast):
         self.assertEqual(tensor.view(1, 6, 2, 1), contig_tensor.view(1, 6, 2, 1))
 
+        # test viewing the same size
+        _tensor = tensor.view([3, 4])
```
this isn't going to catch the issue, as I mentioned above, this has to be done in a non-variable scope.

jerryzh168 replied:
I see, should I still leave them here though

jerryzh168 committed:
Update on "[fix] at::view"


accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

gchanan reviewed (approved):
thanks for fixing this!

jerryzh168 commented:
@pytorchbot retest this please

jerryzh168 commented:
@pytorchbot retest this please

jerryzh168 commented:
@pytorchbot retest this please

jerryzh168 committed:
Update on "[fix] at::view"


accidently calls clone, but what we want is creating an empty tensor and set storage.

Differential Revision: [D16442756](https://our.internmc.facebook.com/intern/diff/D16442756/)

jerryzh168 commented:
@pytorchbot retest this please

jerryzh168 commented:
reexported in: https://github.com/pytorch/pytorch/pull/23452
